cmd/snap-update-ns/utils.go

```diff
@@ -363,6 +363,98 @@ func planWritableMimic(dir string) ([]*Change, error) {
 	return changes, nil
 }
 
+// FatalError is an error that we cannot correct.
+type FatalError struct {
+	error
+}
+
+// execWritableMimic executes the plan for a writable mimic.
+// The result is a transformed mount namespace and a set of fake mount changes
+// that only exist in order to undo the plan.
+//
+// Certain assumptions are made about the plan, it must closely resemble that
+// created by planWritableMimic, in particular the sequence must look like this:
+//
+// - bind a directory aside into safekeeping location
+// - cover the original with tmpfs
+// - bind mount something from safekeeping location to an empty file or
+//   directory in the tmpfs; this step can repeat any number of times
+// - unbind the safekeeping location
+//
+// Apart from merely executing the plan a fake plan is returned for undo. The
+// undo plan skips the following elements as compared to the original plan:
+//
+// - the initial bind mount that constructs the safekeeping directory is gone
+// - the final unmount that removes the safekeeping directory
+// - the source of each of the bind mounts that re-populate tmpfs.
+//
+// In the event of a failure the undo plan is executed and an error is
+// returned. If the undo plan fails the function returns a FatalError as it
+// cannot fix the system from an inconsistent state.
+func execWritableMimic(plan []*Change) ([]*Change, error) {
+	undoChanges := make([]*Change, 0, len(plan)-2)
+	for i, change := range plan {
+		if _, err := changePerform(change); err != nil {
+			recoveryUndoChanges := make([]*Change, 0, len(undoChanges)+1)
+			if i > 0 {
+				// The undo plan doesn't contain the entry for the initial bind
+				// mount of the safe keeping directory but we have already
+				// performed it. For this recovery phase we need to insert that
+				// in front of the undo plan manually.
+				recoveryUndoChanges = append(recoveryUndoChanges, plan[0])
+			}
+			recoveryUndoChanges = append(recoveryUndoChanges, undoChanges...)
+
+			// Drat, we failed! Let's undo everything according to our own undo
+			// plan, by following it in reverse order.
+			for j := len(recoveryUndoChanges) - 1; j >= 0; j-- {
+				recoveryUndoChange := recoveryUndoChanges[j]
+				// All the changes mount something, we need to reverse that.
+				// The "undo plan" is "a plan that can be undone" not "the plan
+				// for how to undo" so we need to flip the actions.
+				recoveryUndoChange.Action = Unmount
+				if _, err2 := changePerform(recoveryUndoChange); err2 != nil {
+					// Drat, we failed when trying to recover from an error.
+					// We cannot do anything at this stage.
+					return nil, &FatalError{error: fmt.Errorf("cannot undo change %q while recovering from earlier error %v: %v", recoveryUndoChange, err, err2)}
```

Comment:
Naive question: I understand we have a chain of changes that depend on predecessors, so if something fails we cannot fully recover anymore; but at the same time (AFAIR) we may have some changes that are independent (e.g. multiple mounts?), wouldn't it make sense to just try undo all mount actions for example to clean as much as possible (and return the error(s) anyway of course)?

Reply:
Also.. Is there a possibility of a transient failure (e.g. a mountpoint being held by a running process?), and a retry later would succeed? Is this something we should be worried about here?

Reply:
There are no running processes that can interfere with things in any way here. First of all we freeze everything so we can be confident that no ongoing mutations occur. Second of all we are constructing a mirror of something that we just mostly read. The final operations that replace the original cannot be stopped by any open files as we simply mount a tmpfs over existing files.

The error path here is really just about apparmor profile being broken. In normal cases where the code is correct it would never happen. If apparmor gets in the way all bets are off (it has the power to prevent us from undoing everything, for example). While I coded this defensively the idea is that it would never fail as a course of normal operation.

Independent changes are handled at a layer above this and this is indeed what we do, if something fails we just carry on with other changes that don't strongly depend on the things we failed with (I think we may even just try all the things, I need to check).

Reply:
Great, thanks for explanation!